File: data_structs/cpp/AlgorithmsAndDataStructs/algorithms/sorting/selection_sort.cpp
//
//  SelectionSort.cpp
//  AlgorithmsAndDataStructs
//
//  Created by Rakesh Patel on 9/21/14.
//  Copyright (c) 2014 Rakesh Patel. All rights reserved.
//

#include "selection_sort.h"
#include <utility>

namespace algo_ns {
    
    void SelectionSort::sort(std::vector<int>& collection) {
        int i=0;
        for (; i < collection.size()-1; ++i) {
            int min = i;
            
            int j = i+1;
            for (; j < collection.size(); ++j) {
                if (collection[j] < collection[min]) {
                    min = j;
                }
            }
            
            if (min != i) {
                int temp = collection[min];
                collection[min] = collection[i];
                collection[i] = temp;
            }
        }
    }
}

File: data_structs/ruby/algorithms/sorting/selection_sort.rb
require_relative '../../utils/utils.rb'

module Algorithms
  class SelectionSort
    include Utils

    def self.sort(collection)
      min = 0

      for i in 0..collection.size-2
        min = i
        for j in i+1..collection.size-1
          min = j if less(collection[j], collection[min])
        end
        collection[i], collection[min] = collection[min], collection[i]
      end
      collection
    end
  end
end
